### DISCUSSION TITLE
Creating list with pybrick blocks

### DISCUSSION THREAD

**Discussion by ogzhnkse** [pybricks/support#1908] [Q&A] [answered]
Hello everyone, I need to create a list but I couldn't find how to add it in block coding. I will create a list and the sensor will write the values it reads every 100ms to the indices in the list. The programme will stop when the length of the list is greater than 6.

**Comment by BertLindeman** [comment]
Maybe a picture can help here:
![image](https://github.com/user-attachments/assets/71dd37e3-f0c5-4ff1-a6bb-da3b5490497b)

**Reply by BertLindeman** [reply]
To define a list variable, you take from "Variables" a block to create the variable and then copy the green block onto it.
Like this: 
![image](https://github.com/user-attachments/assets/4a8e5c79-585b-42f4-a09f-0a8c2570792e)

Bert

**Reply by BertLindeman** [reply]
[EDIT} typo colot =-> color

```python
color_list = []
```

is a good way to define an empty list where you later append sensor data on to.

**Reply by ogzhnkse** [reply]
![Uploading Ekran Resmi 2024-10-30 14.48.24.png…]()
Yes I did it thank you :)

**Reply by ogzhnkse** [reply]
Hi @BertLindeman,
I need help 🙂 

My project is to read the reflection values of six 4 cm wide objects with a 5 cm gap between them using reflected light. While reading each object for 4 cm, it should record the highest value detected and store it in a list. The robot should do this while moving 40 cm along a line. I coded it in Python, but I need the block version.

**Reply by BertLindeman** [reply]
Help I can.

Can you show your python program and where you have a problem doing that in the block program?

Bert

**Reply by ogzhnkse** [reply]
from pybricks.hubs import PrimeHub
from pybricks.parameters import Direction, Port
from pybricks.pupdevices import ColorSensor, Motor
from pybricks.robotics import DriveBase
from pybricks.tools import StopWatch, multitask, run_task, wait

from color_tools import get_hue, get_sat, get_value

# Set up all devices.
prime_hub = PrimeHub()
takip = ColorSensor(Port.E)
tara = ColorSensor(Port.C)
time = StopWatch()
left = Motor(Port.B, Direction.COUNTERCLOCKWISE)
right = Motor(Port.A, Direction.CLOCKWISE)
drive_base = DriveBase(left, right, 62.4, 151)

# Değişkenleri bir liste olarak tanımlayın
boxes = []

# Function to find and print the most frequent values from the list
def most_common(reflections):
    frequency_dict = {}
    for value in reflections:
        if value in frequency_dict:
            frequency_dict[value] += 1
        else:
            frequency_dict[value] = 1
    sorted_values = sorted(frequency_dict.items(), key=lambda item: item[1], reverse=True)
    return sorted_values

# Robot 4 saniye boyunca 100 mm/s hızla ilerleyecek ve her 20 mm'de bir 40 mm boyunca ölçüm yapacak
drive_base.reset()
drive_base.drive(100, 0)  # Set speed to 100 mm/s

time.reset()

total_distance = 0

while time.time() < 4000:  # 4 seconds in milliseconds
    total_distance = drive_base.distance()
    if total_distance % 20 == 0 and total_distance > 0:
        # 40 mm boyunca common reflection hesapla
        drive_base.stop()
        reflections = []
        drive_base.reset()
        drive_base.drive(100, 0)
        while drive_base.distance() < 40:
            reflection = tara.reflection()
            if reflection > 2:  # Record values greater than 2
                reflections.append(reflection)
            wait(10)  # Wait for 10 milliseconds between readings
        if reflections:
            common_reflection = most_common(reflections)[0][0]
        else:
            common_reflection = 0

        # Box değerlerine common_reflection değerini ekle
        boxes.append(common_reflection)

        print(f"Box{len(boxes)} değer: {common_reflection}")
    
    wait(10)  # Small delay to ensure distance is updated correctly

# Stop the drive base after 4 seconds
drive_base.stop()

# Function to find the most frequent values and print them
def find_most_frequent(reflections):
    if reflections:
        most_common_values = most_common(reflections)
        for value, frequency in most_common_values:
            print(f"Value: {value}, Frequency: {frequency}")

# Main function to execute the recording and analysis
def main():
    find_most_frequent(boxes)

# Run the main function
main()

**Reply by ogzhnkse** [reply]
<img width="449" alt="Ekran Resmi 2024-11-08 11 40 57" src="https://github.com/user-attachments/assets/31f8f0dc-1269-420a-8f23-c1110a4b6445">
<img width="368" alt="Ekran Resmi 2024-11-08 11 49 04" src="https://github.com/user-attachments/assets/7388b681-294a-4fff-8b49-3e298c0df3a5">

**Reply by ogzhnkse** [reply]
My problem is that, previously I asked how to create a list, and now I'm trying to add the most frequently occurring value of the readings to the list. I wrote a function for that, but I'm having trouble adding the result of the function to the list.

**Reply by BertLindeman** [reply]
Something went wrong with the upload of your two pictures.
They look to me as:
```
Uploading Ekran Resmi 2024-11-08 11.40.57.png…

Uploading Ekran Resmi 2024-11-08 11.49.04.png…
````
Can you re-try the upload? 
Maybe wait a bit after the upload until it's ready? (Just guessing though).

**Reply by BertLindeman** [reply]
Maybe this helps.
To add a value to a list, you can use this block.
Use the arrow to select "append..."
![pybricks_program_2024-11-08T09_11_33](https://github.com/user-attachments/assets/a9a24f6f-dbce-4881-b493-799943bb95dc)

**Reply by ogzhnkse** [reply]
It was caused by the WiFi.

**Reply by BertLindeman** [reply]
Thank you. 
Does this hint help:
![image](https://github.com/user-attachments/assets/ccf908ee-3cd6-48ec-9765-8eedcec13863)

[EDIT] Have to go out for the day, so no more time to help you at the moment.

**Reply by ogzhnkse** [reply, marked as answer]
Thank you very  much, I'll try it and let you know. Thank you again for your help. Wishing you a good day!
